### PR TITLE
WTrackMenu: Add ability to select loaded track in library

### PR DESCRIPTION
**Pull request by ninomp**
This PR allows user to highlight (select) track that is currently loaded to deck or sampler in library. This works across library features (Tracks, AutoDJ, Playlists, Crates).

### CODE REVIEW

**Comment by daschuer** [comment]
Nice. Thank you.

**Comment by Russe** [comment]
This is a very nice feature, thank you.
I tested this PR with Utbuntu and it works like a charm.

One questions, do you think it's also possible to mark or highlight (not select) loaded tracks automatically in library?
This could be an additional feature.

File: src/widget/wtracktableview.h [hunk 1]
```diff
@@ -69,6 +69,10 @@ class WTrackTableView : public WLibraryTableView {
     bool slotRestoreCurrentViewState() {
         return restoreCurrentViewState();
     };
+    void slotSelectTrack(const TrackId& trackId) {
```
Comment (uklotzde) [inline src/widget/wtracktableview.h:72]: Please move the implementation into the .cpp file. I suggest to define only trivial, single-line functions inline.

**Review by uklotzde — APPROVED** [review]
One minor nitpick, otherwise fine.

**Comment by daschuer** [comment]
Thank you for the feature. It works OK. 
Can you also scroll to the selected track if it is already selected. 
It would be also nice, if the current would be selected in case you have searched a similar track. 
The feature dos not work in case AutoDJ is selected. In this case the cover art is shown regardless of the track in in the AutoDJ List or not.
My expectation was that this feature works for the currently selected library table. Helpfull to see the current track in the context of the others in a crate or library.

File: src/widget/wtracktableview.cpp [hunk 2]
```diff
@@ -1011,6 +1011,11 @@ void WTrackTableView::slotAddToAutoDJReplace() {
     addToAutoDJ(PlaylistDAO::AutoDJSendLoc::REPLACE);
 }
 
+void WTrackTableView::slotSelectTrack(const TrackId& trackId) {
+    setCurrentTrackId(trackId);
```
Comment (daschuer) [inline src/widget/wtracktableview.cpp:1015]: I think here is a check missing if the track is in that Table at all.

**Review by daschuer — CHANGES_REQUESTED** [review]

**Comment by ronso0** [comment]
@ninomp Please reword the last two commits to provide meaningful messages, rebase onto the commit before [b80c7b3](https://github.com/mixxxdj/mixxx/pull/4688/commits/b80c7b3354c1cfdffce294230a97108807a5dc7f) and force-push (no one has reviewed them so far so it should be okay)

**Comment by ronso0** [comment]
> @ninomp Please reword the last two commits to provide meaningful messages, rebase onto the commit before [b80c7b3](https://github.com/mixxxdj/mixxx/pull/4688/commits/b80c7b3354c1cfdffce294230a97108807a5dc7f) and force-push (no one has reviewed them so far so it should be okay)

Sorry I didn't notice those were merges from main. Why didn't you use the usual `merge upstream/main`?

File: src/widget/wtrackproperty.cpp [hunk 3]
```diff
@@ -21,7 +21,8 @@ constexpr WTrackMenu::Features kTrackMenuFeatures =
         WTrackMenu::Feature::RemoveFromDisk |
         WTrackMenu::Feature::FileBrowser |
         WTrackMenu::Feature::Properties |
-        WTrackMenu::Feature::UpdateReplayGainFromPregain;
+        WTrackMenu::Feature::UpdateReplayGainFromPregain |
+        WTrackMenu::Feature::SelectInLibrary;
```
Comment (ronso0) [inline src/widget/wtrackproperty.cpp:25]: This also needs to be added to WTrackWidgetGroup, used for the track color bar in the decks

File: src/widget/wtracktableview.cpp [hunk 4]
```diff
@@ -1034,6 +1039,11 @@ void WTrackTableView::slotAddToAutoDJReplace() {
     addToAutoDJ(PlaylistDAO::AutoDJSendLoc::REPLACE);
 }
 
+void WTrackTableView::slotSelectTrack(const TrackId& trackId) {
+    setCurrentTrackId(trackId, 0, true);
+    setSelectedTracks({trackId});
```
Comment (ronso0) [inline src/widget/wtracktableview.cpp:1044]:
```suggestion
    if (setCurrentTrackId(trackId, 0, true)) {
        setSelectedTracks({trackId});
    }
```

**Review by ronso0 — CHANGES_REQUESTED** [review]

File: src/library/library.cpp [hunk 5]
```diff
@@ -400,6 +400,7 @@ void Library::bindLibraryWidget(
             &Library::restoreModelState,
             pTrackTableView,
             &WTrackTableView::slotRestoreCurrentViewState);
+    connect(this, &Library::selectTrack, pTrackTableView, &WTrackTableView::slotSelectTrack);
```
Comment (ronso0) [inline src/library/library.cpp:403]: for readability, please break this line like the surrounding connections.

**Review by ronso0 — CHANGES_REQUESTED** [review]

**Comment by ronso0** [comment]
re @daschuer's observation
> The feature dos not work in case AutoDJ is selected. In this case the cover art is shown regardless of the track in in the AutoDJ List or not.

This happens when clicking _any_ sidebar root item (feature root), except Tracks and History.
With those, the table model is simply not "shown" when the view is selected, so the `selectTrack` request goes to the previously selected WTrackTableView. Btw that also clears the previous track selection there :|

@ninomp
Since I messed with the library a few times already I was curious and took a look.
For features where the root view has a track view (AutoDJ, Analyze, Recording), I managed to fix it by adding `emit showTrackModel( [trackModel] );` to `[Feature]::activate`. If you mind that approach take a look at 8cfd397773
For other root views _without_ a tracks table (Crates, Playlists, Browse, some external libraries?), no idea..

Maybe the `selectTrack` signal should rather be connected in `Library::addFeature`, and then each feature needs a `slotSelectTrack` slot? Maybe also make use of `WLibrary::getActiveView()`, or add `bool `WLibrary::isCurrentViewTracksTable()`? Possible, if features store the WLibrary pointer in `feature::bindLibraryWidget`.
https://github.com/mixxxdj/mixxx/blob/6f8676073e1689a1cc749a701d909570d0cafb0e/src/widget/wlibrary.cpp#L91-L93

Sorry, I'm aware this may be overwhelming since your current minimal approach is actually beautifully simple and seemed to work fine, too.

one more thing... :grimacing: 
I'd actually expect the "Select track" action to be greyed out if the current library view is not a WTrackTableView, though it is probably complete overkill to make the decks' track menu aware of the tracks table (if possible at all).

**Comment by daschuer** [comment]
https://github.com/mixxxdj/mixxx/pull/4740 has been merged, so this can be closed. @ninomp Thank you for starting this nice feature.

**Pull request closed without merging**